xds/internal/xdslbregistry/converter.go

```diff
@@ -0,0 +1,288 @@
+		wrrLBConfig.EnableOOBLoadReport = enableOOBLoadReportCfg.GetValue()
+	}
+	if oobReportingPeriodCfg := cswrrProto.GetOobReportingPeriod(); oobReportingPeriodCfg != nil {
+		wrrLBConfig.OOBReportingPeriod = internalserviceconfig.Duration(oobReportingPeriodCfg.AsDuration())
+	}
+	if blackoutPeriodCfg := cswrrProto.GetBlackoutPeriod(); blackoutPeriodCfg != nil {
+		wrrLBConfig.BlackoutPeriod = internalserviceconfig.Duration(blackoutPeriodCfg.AsDuration())
+	}
+	if weightExpirationPeriodCfg := cswrrProto.GetBlackoutPeriod(); weightExpirationPeriodCfg != nil {
+		wrrLBConfig.WeightExpirationPeriod = internalserviceconfig.Duration(weightExpirationPeriodCfg.AsDuration())
+	}
+	if weightUpdatePeriodCfg := cswrrProto.GetWeightUpdatePeriod(); weightUpdatePeriodCfg != nil {
+		wrrLBConfig.WeightUpdatePeriod = internalserviceconfig.Duration(weightUpdatePeriodCfg.AsDuration())
+	}
+	if errorUtilizationPenaltyCfg := cswrrProto.GetErrorUtilizationPenalty(); errorUtilizationPenaltyCfg != nil {
+		wrrLBConfig.ErrorUtilizationPenalty = float64(errorUtilizationPenaltyCfg.GetValue())
+	}
+
+	lbCfgJSON, err := json.Marshal(wrrLBConfig)
+	if err != nil {
+		return nil, false, fmt.Errorf("error marshaling JSON for type %T: %v", wrrLBConfig, err)
+	}
+	return makeBalancerConfigJSON(weightedroundrobin.Name, lbCfgJSON), false, nil
+}
+
+type v1TypedStructConverter struct{}
+
+func (v1tsc *v1TypedStructConverter) convertToServiceConfig(rawProto []byte, depth int) (json.RawMessage, bool, error) {
+	tsProto := &v1xdsudpatypepb.TypedStruct{}
+	if err := proto.Unmarshal(rawProto, tsProto); err != nil {
+		return nil, false, fmt.Errorf("failed to unmarshal resource: %v", err)
+	}
+	return convertCustomPolicy(tsProto.GetTypeUrl(), tsProto.GetValue())
+}
+
+type v3TypedStructConverter struct{}
+
+func (v3tsc *v3TypedStructConverter) convertToServiceConfig(rawProto []byte, depth int) (json.RawMessage, bool, error) {
+	tsProto := &v3xdsxdstypepb.TypedStruct{}
+	if err := proto.Unmarshal(rawProto, tsProto); err != nil {
+		return nil, false, fmt.Errorf("failed to unmarshal resource: %v", err)
+	}
+	return convertCustomPolicy(tsProto.GetTypeUrl(), tsProto.GetValue())
+}
+
+// convertCustomPolicy attempts to prepare json configuration for a custom lb
+// proto, which specifies the gRPC balancer type and configuration. Returns the
+// converted json, a bool representing whether the caller should continue to the
+// next policy, which is true if the gRPC Balancer registry does not contain
+// that balancer type, and an error which should cause caller to error if error
+// converting.
+func convertCustomPolicy(typeURL string, s *structpb.Struct) (json.RawMessage, bool, error) {
+	// The gRPC policy name will be the "type name" part of the value of the
+	// type_url field in the TypedStruct. We get this by using the part after
+	// the last / character. Can assume a valid type_url from the control plane.
+	urls := strings.Split(typeURL, "/")
+	name := urls[len(urls)-1]
```

Comment:
Nit, but slightly more efficient as it doesn't require allocating a slice:

```go
pos := strings.LastIndex(typeURL, "/")
name := typeURL[pos+1:]
```

Reply:
Done.